### PR TITLE
Generate reduce

### PR DESCRIPTION
**Pull request by SohamTamba**
`mapreduce` expects other processors to have access to data.
Otherwise, it executes sequentially.

I used `@threads` instead to run it in parallel.

*cc* @somil55

### CODE REVIEW

**Comment by somil55** [comment]
@SohamTamba How about doing it in a distributed way? You might get a better performance using `pmap` and then reducing it at the main worker

**Comment by SohamTamba** [comment]
This issue with `pmap` is it will use `Reps*nv(g)` memory.

I tried using `@distributed` (with 4 workers) but it does not seem to improve performance.

```
function generate_min_set(g::AbstractGraph{T}, gen_func::Function, Reps::Integer) where T<:Integer

    min_set::Vector{T} = @distributed ((x::Vector{T}, y::Vector{T})->length(x)<length(y) ? x : y) for i in 1:Reps
        gen_func(g)
    end

    return min_set
end
```

**Comment by somil55** [comment]
ok!

**Comment by somil55** [comment]
@sbromberger Can you please review it?

**Comment by SohamTamba** [comment]
I'll make the change in `generate_max_set` along with the next Hueristic after this and `VertexCover` is accepted.
I'd prefer to merge as many PR's as possible before GSoC ends.

**Comment by SohamTamba** [comment]
I coded a distributed implementation using `remotecall_fetch`.

Foe Ego Twitter graph which has 80,000 vertices, the threaded implenetation has the same run time as the distributed implementation when `Reps = 1000`.

The run-time of the distributed increases because the graph has to be copied in every process.

Benefit of distributed: The code will be more memory efficient since the processes dont share memory and cache

Disadvantage: Each process must copy the graph, increasing run-time.

I think the threaded implementation would be better because it seems unlikely for `Reps` to be more than `100`.

Let me know if you want me to PR the distributed implementation.

**Comment by somil55** [comment]
@SohamTamba In that case, threaded version seems more practical. What do you think @sbromberger?

**Comment by sbromberger** [comment]
Why not both with a kwarg in a wrapper function? `parallel=: threads` or something?

**Comment by SohamTamba** [comment]
That makes sense.

`@threads` is more suitable for this algorithm because it performs fast operations while iterating over the  graph only once.

`remotecall_fetch` might be faster for other algorithms. This is the case for centrality measures. The distributed implementation  already present is faster than the multi-threaded version I implemented.

**Comment by SohamTamba** [comment]
@sbromberger 
I added a multithreaded and distributed implementation and a wrapper function.

Btw it turns out `@distributed` implementation is faster than the one with the remote calls.
I just so happen that for `Reps = 20`, the runtime was `1.2 s` and a single sequential call had a runtime of `0.06s`.
However, `mapeduce` did not run in parallel.

File: src/utils.jl [hunk 1]
```diff
@@ -57,13 +57,60 @@ function insorted(item, collection)
     @inbounds return (index <= length(collection) && collection[index] == item)
 end
 
+
 """
-    generate_min_set(g, gen_func, Reps)
+    generate_min_set(g, gen_func, Reps; parallel=:threads)
```
Comment (sbromberger) [inline src/utils.jl]: Why is `Reps` capitalized? AFAIK we don't do that anywhere else for variables greater than 1 or 2 chars.

Comment (sbromberger) [reply src/utils.jl]: Also, with the new `Parallel` module, this should probably go into `Parallel/utils.jl`.

File: src/utils.jl [hunk 2]
```diff
@@ -61,8 +61,29 @@ end
     generate_min_set(g, gen_func, Reps)
 Generate a vector `Reps` times using `gen_func(g)` and return the vector with the least elements.
 """
-generate_min_set(g::AbstractGraph{T}, gen_func, Reps::Integer) where T<: Integer =
-mapreduce(gen_func, (x, y)->length(x)<length(y) ? x : y, Iterators.repeated(g, Reps))
+function generate_min_set(g::AbstractGraph{T}, gen_func::Function, Reps::Integer) where T<: Integer
```
[the remark this answers is not in the document]
Comment (SohamTamba) [reply src/utils.jl]: That is the convention I had noticed online.
I can change it to `reps` instead.

File: src/utils.jl (hunk 1 above)
Comment (SohamTamba) [reply src/utils.jl]: I've included it in the parallel module.
Should I replace the implementation using `mapreduce` from the sequential model?

**Comment by SohamTamba** [comment]
I was just thinking,
instead of having 2 functions for `min_set` and `max_set`, how about I use `Base.ordering` and create only one function?

**Comment by SohamTamba** [comment]
On another note, I noticed I tested `max_set` twice instead of `min_set` and `max_set` once each.

Should I create a separate PR to fix it?

File: src/Parallel/utils.jl [hunk 3]
```diff
@@ -0,0 +1,53 @@
+"""
+    generate_min_set(g, gen_func, Reps; parallel=:threads)
+
+Generate a vector `Reps` times using `gen_func(g)` and return the vector with the least elements.
```
[the remark this answers is not in the document]
Comment (SohamTamba) [reply src/Parallel/utils.jl]: Yeah.
I thought I had done that.

Must be an issue from rebasing.

**Pull request merged**